record/envelope.go

```diff
@@ -0,0 +1,196 @@
+package record
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"time"
+
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	pb "github.com/libp2p/go-libp2p-core/record/pb"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/multiformats/go-varint"
+)
+
+// SignedEnvelope contains an arbitrary []byte payload, signed by a libp2p peer.
+//
+// Envelopes are signed in the context of a particular "domain", which is a
+// string specified when creating and verifying the envelope. You must know the
+// domain string used to produce the envelope in order to verify the signature
+// and access the payload.
+type SignedEnvelope struct {
+	// The public key that can be used to verify the signature and derive the peer id of the signer.
+	PublicKey crypto.PubKey
+
+	// A binary identifier that indicates what kind of data is contained in the payload.
+	// TODO(yusef): enforce multicodec prefix
+	PayloadType []byte
+
+	// A monotonically-increasing sequence counter for ordering SignedEnvelopes in time.
+	Seq uint64
```

Comment:
Now we have a sequence number here and inside the peer record?

Reply:
Let's keep it inside the peer record.

Reply:
hehe, I moved it over based on your comment here, https://github.com/libp2p/go-libp2p-core/pull/73#discussion_r361714748, but I'm happy to move it back.

One thing about having the `Seq` field in the `PeerRecord` that's a bit awkward is that, if we remove the `NewPeerRecord` constructor (which I agree is a good idea), then people constructing a `PeerRecord` struct directly will need to explicitly set the `Seq` number, e.g.:

```go
rec := &PeerRecord{PeerID: myId, Addrs: myAddrs, Seq: generateSeqNo()}
```

Or I suppose we could set it for you when you marshal the record if `Seq == 0`, but it feels wrong for a marshal function to alter the thing it's marshaling.